pkg/parse/json/json_test.go

```diff
@@ -0,0 +1,552 @@
+			t.Logf("Validating %s", eventBindings[i])
+
+			// Remove spacing to normalize both values
+			actualBuffer := new(bytes.Buffer)
+			expectedBuffer := new(bytes.Buffer)
+			if err := gojson.Compact(actualBuffer, []byte(actualValues[i])); err != nil {
+				t.Errorf("Unable to compact actual bytes for field %s", eventBindings[i])
+				continue
+			}
+			if err := gojson.Compact(expectedBuffer, []byte(expectedValues[i])); err != nil {
+				t.Errorf("Unable to compact expected bytes for field %s", eventBindings[i])
+				continue
+			}
+			// Grab token to determine value type
+			expectedBufferCopy := bytes.NewBuffer(expectedBuffer.Bytes())
+			decoder := gojson.NewDecoder(expectedBufferCopy)
+			token, err := decoder.Token()
+			if err != nil {
+				t.Errorf("Error getting token from expected bytes for field %s", eventBindings[i])
+				continue
+			}
+			_, ok := token.(gojson.Delim)
+			if ok { // Compare json structures
+				var expected, actual interface{}
+				err := gojson.Unmarshal(expectedBuffer.Bytes(), &expected)
+				if err != nil {
+					t.Fatal("Failed to unmarshal expected value into map for comparison\nError:", err)
+				}
+				err = gojson.Unmarshal(actualBuffer.Bytes(), &actual)
+				if err != nil {
+					t.Fatal("Failed to unmarshal actual value into map for comparison\nError:", err)
+				}
+				if !reflect.DeepEqual(expected, actual) {
+					t.Errorf("Event binding to field %s did not yield correct value. \nExpected: %v\nActual: %v", eventBindings[i], expected, actual)
+				}
+			} else { // Compare json fields
+				expected, actual := expectedBuffer.String(), actualBuffer.String()
+				if expected != actual {
+					t.Errorf("Event binding to field %s did not yield correct value. \nExpected: %v\nActual: %v", eventBindings[i], expected, actual)
+				}
+			}
+		}
+	}
+}
+
+// Digs into map[string]interface{} recursively whenever the value is a json object
+// Returns all possible event bindings and corresponding expected values
+// expectedBytes have been marshalled and have a potential random order compared to the initial map
+func eventBindingDigger(t *testing.T, m map[string]interface{}, location string) (eventBindings []string, expectedValues []string) {
+	b, _ := gojson.Marshal(m)
+	eventBindings = append(eventBindings, parse.Interpolate(location))
+	expectedValues = append(expectedValues, string(b))
+	for field, value := range m {
+		currentLocation := fmt.Sprintf("%s.%s", location, field)
+		b, err := gojson.Marshal(value)
+		if err != nil {
+			t.Fatalf("Failed to marshal value %v", value)
+		}
+		eventBindings = append(eventBindings, parse.Interpolate(currentLocation))
+		expectedValues = append(expectedValues, string(b))
+		nestedMap, ok := value.(map[string]interface{})
+		if ok {
+			nestedBindings, nestedValues := eventBindingDigger(t, nestedMap, currentLocation)
+			eventBindings, expectedValues = append(eventBindings, nestedBindings...), append(expectedValues, nestedValues...)
+		}
+	}
+	return
+}
```

Comment:
im really glad to see the test logic getting a lot of love!

at the moment if I was reading these tests (and especially updating them) I'd have a bit of trouble understanding exactly what was going on b/c there is so much logic in the test function - and I'd be nervous to trust it b/c it itself is untested code

What do you think about breaking up the test into multiple functions, and actually having unit tests for those (including `eventBindingDigger`)?

Reply:
Fair point.

Reply:
> However, if we were include many example payloads, manually enumerating all of their potential mappings might become cumbersome

I see what you mean! At the moment we only have 2, so I feel like manually including the mappings might be feasible? We could revisit as we want to add more.